### ISSUE TITLE
Is there any possibility to put arguments in the builder pattern.

### ISSUE THREAD

**Issue by portlek** [JorelAli/CommandAPI#125]
I mean, I can show an example actually, for example, I made that design for any Java project actually but it would be good if we can implement it into Bukkit;
(example from [commands](https://github.com/portlek/commands) )
```java
// For the bukkit projects.
final class BukkitCmdTest {

    @NotNull
    final JavaPlugin plugin;

    BukkitCmdTest(@NotNull final JavaPlugin plugin) {
        this.plugin = plugin;
    }

    void creation() {
        final BukkitCmdRegistry registry = new BukkitCmdRegistry(this.plugin);
        final BukkitCmdBasic command = new BukkitCmdBasic("test-command", "test-aliases")
            .permission("plugin.test-command.main")
            .guard(context ->
                true)
            .execute(context -> {
                // executes /test-command
            })
            .createSub("help", subCmd -> subCmd
                .permission("plugin.test-command.help")
                .executePrevious(true))
            .createSub("giveseed", seedsub -> seedsub
                .permission("plugin.test-command.giveseed")
                .executePrevious(true)
                .createSub("seeds", seedssub -> seedssub
                    .type(ArgType.text("<seed-name>"))
                    .executePrevious(true)
                    .createSub("player-argument", playersub -> playersub
                        .type(BukkitArgType.players())
                        .execute(context -> {
                            // executes /test-command giveseed <seed-name> <player> [amount]
                            final int size = context.args().size();
                            final int amount;
                            final Player player;
                            if (size == 4) {
                                amount = context.arg().asNumber(Integer.class, "[Prefix] > Input a number!");
//                                player = context.previous().asPlayer("[Prefix] > Player not found!");
                            } else {
                                amount = 1;
//                                player = context.arg().asPlayer("[Prefix] > Player not found!");
                            }

                        })
                        .createSub("amount-argument", amountsub -> amountsub
                            .type(ArgType.integer("[amount]"))
                            .executePrevious(true)))))
            .createSub("reload", reloadsub -> reloadsub
                .permission("plugin.test-command.reload")
                .execute(context -> {
                    // executes /test-command reload
                }))
            .createSub("list", subCmd -> subCmd
                .executePrevious(true)
                .createSub("page-argument", pagesub -> pagesub
                    .type(ArgType.integer("[page]"))

                    .execute(context -> {
                        // executes /test-command list [page]
                    })))
            .createSub("perm", subCmd -> subCmd
                .permission("plugin.test-command.perm")
                .executePrevious(true)
                .createSub("list", listsub -> listsub
                    .permission("plugin.test-command.perm.list")
                    .executePrevious(true)
                    .createSub("page-argument", pagesub -> pagesub
                        .type(ArgType.integer("[page]"))
                        .execute(context -> {
                            // executes test-command perm list [page]
                        })))
                .createSub("add", addsub -> addsub
                    .permission("plugin.test-command.perm.add")
                    .executePrevious(true)
                    .createSub("player-argument", playersub -> playersub
                        .type(BukkitArgType.players())
                        .execute(context -> {
                            // executes test-command perm add <player>
                        })))
                .createSub("remove", removesub -> removesub
                    .permission("plugin.test-command.perm.remove")
                    .executePrevious(true)
                    .createSub("player-argument", playersub -> playersub
                        .type(BukkitArgType.players())
                        .execute(context -> {
                            // executes test-command perm remove <player>
                        }))));
        registry.register(command);
    }

}
```

**Comment by portlek** [comment]
I know it's possible with some tricky ways but, is there any simple way to do that?

**Comment by JorelAli** [comment]
Hi,
So I see where you're coming from - why the builder pattern is suitable for this sort of thing. Since it's late at the moment, I'll try and keep this brief.

There are multiple different ways to approach command registration patterns. A few of the most common ones are:
- A builder pattern, such as [Mojang's brigadier](https://github.com/Mojang/brigadier)
- An annotation-based approach, such as [ACF](https://github.com/aikar/commands)
- Whatever pattern Bukkit uses where you just override some `onCommand()` method.

The CommandAPI was originally designed to mimic Bukkit's format - a simple declarative approach for arguments, followed by a "simple" builder-like approach for commands where the main "command execute" is the main part of the code. This design was chosen since it makes it somewhat simpler to switch to for developers that are already comfortable with the standard Bukkit system.

I personally found that the builder pattern quickly becomes very complicated. For example, Mojang's brigadier's builder system tends to be very verbose and hard to understand when looking at it from a glance. The CommandAPI's design where arguments are declared first, then commands are defined later is a little more intuitive and it allows developers to easily comment their code and segment it into the two main parts: arguments (type checking) and commands (execution).

-----

As you can see from your example, it's very verbose. It's one giant expression to register one very complex command. 

```
command
  subcommand
    subcommand
      subcommand
  subcommand
  subcommand
  ...
```

Without indentation, it's near-impossible to comprehend. The CommandAPI's design lets you easily identify parts of the command since code generally follows this structure:

```
arguments
command

arguments
subcommand

arguments
command

...
```

-----

For these main reasons, I don't plan to change the pattern for command and argument declaration in the CommandAPI. If you want to write a separate wrapper that adds the builder pattern for arguments, by all means feel free, but this is not a suggestion that I ever plan to consider due to it failing to meet one of the the basic specifications of the project _(Being somewhat similar to the Bukkit API's command declaration)_.

**Comment by JorelAli** [comment]
Today is your lucky day.

**Comment by JorelAli** [comment]
Implemented subcommands in commits 23ba9938213a3b41dbb33b6ca70e135f855fbac3 .. d21da5bb0b3b2ac738613dcc82914fc07e991535, which can be used with a builder-pattern style of programming. This has been scheduled for v5.0 release. Extra kudos go to _Minenash_ and _Blocky Borzan_ for helping to make this feature a reality!
